File: typescript/src/examples/ex10-cost-helper.ts
/**
 * Example 10 — Cost/billing helper CLI.
 *
 * Local-only arithmetic over the documented pricing table (1 credit = 1 cent
 * = $0.01 exactly) via the shared `cost.ts` estimator. Subcommands:
 *
 *   predict   --width --height --trajectory --cua --prompt-chars
 *   session   --steps + the predict flags (create is 10 cr, surcharge-free)
 *   ground    --width --height                       (3 cr, +1 if HD)
 *   run       --steps --cua                          (5 cr/step, 8 on v1)
 *   workflow  --task-steps --cua                     (control-flow steps FREE)
 *   machine   --os|--windows --hours --stopped-hours --snapshots
 *   plan      --file batch.json   (totals a JSON batch of the above)
 *
 * Surcharges itemized: +2 cr per trajectory screenshot, +1 cr per HD image
 * (STRICTLY width > 1280 OR height > 720 — exactly 1280x720 is NOT HD,
 * applied to the current shot AND each trajectory shot), +3 cr per request on
 * the v1 engine, +1 cr when system_prompt exceeds 500 chars (exactly 500 is
 * free). Machine runtime is metered per minute and ROUNDED DOWN.
 *
 * Estimated cost of running this example: 0 credits ($0.00) — it never calls
 * the API. Every report also reminds you that sandbox keys never bill and
 * that charges are debited up front and auto-refunded on failure.
 *
 * Run it:
 *   npx tsx src/examples/ex10-cost-helper.ts predict --width 1280 --height 720
 *   npx tsx src/examples/ex10-cost-helper.ts machine --windows --hours 2 --snapshots 1
 *   npx tsx src/examples/ex10-cost-helper.ts plan --file ./batch.json
 */
import { readFileSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { parseArgs } from 'node:util';

import {
  PRICING,
  SYSTEM_PROMPT_FREE_CHARS,
  creditsToUsd,
  estimateGroundCredits,
  estimateMachineRuntimeCredits,
  estimatePredictCredits,
  estimateRunCredits,
  estimateSessionCreateCredits,
  estimateSessionPredictCredits,
  estimateWorkflowRunCredits,
  formatEstimate,
  isHdImage,
  snapshotCredits,
  type EstimateLineItem,
} from '../coasty/cost.js';
import { isSandboxKey } from '../coasty/env.js';
import { type CuaVersion, type OsType } from '../coasty/types.js';

// ---------------------------------------------------------------------------
// Report shape
// ---------------------------------------------------------------------------

export interface CostReport {
  title: string;
  items: EstimateLineItem[];
  totalCredits: number;
  /** Dollars, exact: credits / 100. */
  totalUsd: number;
  notes: string[];
}

export const STANDARD_NOTES: readonly string[] = [
  'sandbox keys (sk-coasty-test-*) never bill: every total is $0.00 in test mode',
  'charges are debited up front and auto-refunded on failure (incl. snapshots)',
];

export class UsageError extends Error {
  constructor(message: string) {
    super(`${message}\n\n${USAGE}`);
    this.name = 'UsageError';
  }
}

function finalize(title: string, items: EstimateLineItem[], notes: string[] = []): CostReport {
  const totalCredits = items.reduce((sum, item) => sum + item.credits, 0);
  return {
    title,
    items,
    totalCredits,
    totalUsd: creditsToUsd(totalCredits),
    notes: [...notes, ...STANDARD_NOTES],
  };
}

// ---------------------------------------------------------------------------
// Per-operation estimators (each returns an ITEMIZED report)
// ---------------------------------------------------------------------------

export interface InferenceArgs {
  width: number;
  height: number;
  trajectory: number;
  cuaVersion: CuaVersion;
  promptChars: number;
}

export const INFERENCE_DEFAULTS: InferenceArgs = {
  width: 1920,
  height: 1080,
  trajectory: 0,
  cuaVersion: 'v3',
  promptChars: 0,
};

function inferenceSurchargeItems(args: InferenceArgs, perRequest: number): EstimateLineItem[] {
  const items: EstimateLineItem[] = [];
  if (args.trajectory > 0) {
    items.push({
      label: `trajectory screenshots (${String(args.trajectory)} x ${String(PRICING.trajectoryScreenshot)} cr)`,
      credits: perRequest * args.trajectory * PRICING.trajectoryScreenshot,
    });
  }
  if (isHdImage(args.width, args.height)) {
    items.push({
      label: `HD surcharge (${String(args.width)}x${String(args.height)} > 1280x720 strict; current + trajectory shots)`,
      credits: perRequest * (1 + args.trajectory) * PRICING.hdImage,
    });
  }
  if (args.cuaVersion === 'v1') {
    items.push({ label: 'v1 engine surcharge', credits: perRequest * PRICING.v1Engine });
  }
  if (args.promptChars > SYSTEM_PROMPT_FREE_CHARS) {
    items.push({
      label: `system_prompt > ${String(SYSTEM_PROMPT_FREE_CHARS)} chars (exactly ${String(SYSTEM_PROMPT_FREE_CHARS)} is free)`,
      credits: perRequest * PRICING.longSystemPrompt,
    });
  }
  return items;
}

export function predictReport(partial: Partial<InferenceArgs> = {}): CostReport {
  const args = { ...INFERENCE_DEFAULTS, ...partial };
  const items: EstimateLineItem[] = [
    { label: 'POST /predict base', credits: PRICING.predict },
    ...inferenceSurchargeItems(args, 1),
  ];
  const report = finalize('predict (one request)', items);
  // Known-answer guard: the itemization must equal the canonical estimator.
  assertTotal(report, estimatePredictCredits(toCostOptions(args)));
  return report;
}

export interface SessionArgs extends InferenceArgs {
  steps: number;
}

export function sessionReport(partial: Partial<SessionArgs> = {}): CostReport {
  const args = { ...INFERENCE_DEFAULTS, steps: 1, ...partial };
  const items: EstimateLineItem[] = [
    {
      label: 'POST /sessions create (one-time, NO surcharges)',
      credits: estimateSessionCreateCredits(),
    },
    {
      label: `session predicts (${String(args.steps)} x ${String(PRICING.sessionPredict)} cr base)`,
      credits: args.steps * PRICING.sessionPredict,
    },
    ...inferenceSurchargeItems(args, args.steps),
  ];
  const report = finalize(`session (create + ${String(args.steps)} steps)`, items, [
    'session reset/get/list/delete are free',
  ]);
  assertTotal(
    report,
    estimateSessionCreateCredits() +
      args.steps * estimateSessionPredictCredits(toCostOptions(args)),
  );
  return report;
}

export function groundReport(
  partial: Partial<Pick<InferenceArgs, 'width' | 'height'>> = {},
): CostReport {
  const args = { ...INFERENCE_DEFAULTS, ...partial };
  const items: EstimateLineItem[] = [{ label: 'POST /ground base', credits: PRICING.ground }];
  if (isHdImage(args.width, args.height)) {
    items.push({
      label: `HD surcharge (${String(args.width)}x${String(args.height)})`,
      credits: PRICING.hdImage,
    });
  }
  const report = finalize('ground (one request)', items);
  assertTotal(
    report,
    estimateGroundCredits({ screenWidth: args.width, screenHeight: args.height }),
  );
  return report;
}

export interface RunArgs {
  steps: number;
  cuaVersion: CuaVersion;
}

export function runReport(partial: Partial<RunArgs> = {}): CostReport {
  const args: RunArgs = { steps: 10, cuaVersion: 'v3', ...partial };
  const perStep = args.cuaVersion === 'v1' ? PRICING.runStepV1 : PRICING.runStepV3V4;
  const report = finalize(
    `run (${String(args.steps)} agent steps, ${args.cuaVersion})`,
    [
      {
        label: `run steps (${String(args.steps)} x ${String(perStep)} cr on ${args.cuaVersion})`,
        credits: args.steps * perStep,
      },
    ],
    [
      'run steps have NO other surcharges (HD/trajectory/prompt do not apply)',
      'starting a run requires the wallet to cover at least one step',
    ],
  );
  assertTotal(report, estimateRunCredits({ steps: args.steps, cuaVersion: args.cuaVersion }));
  return report;
}

export function workflowReport(
  partial: Partial<{ taskSteps: number; cuaVersion: CuaVersion }> = {},
): CostReport {
  const args = { taskSteps: 10, cuaVersion: 'v3' as CuaVersion, ...partial };
  const perStep = args.cuaVersion === 'v1' ? PRICING.runStepV1 : PRICING.runStepV3V4;
  const report = finalize(
    `workflow run (${String(args.taskSteps)} task-step executions, ${args.cuaVersion})`,
    [
      {
        label: `task steps (${String(args.taskSteps)} x ${String(perStep)} cr)`,
        credits: args.taskSteps * perStep,
      },
      { label: 'control-flow steps (assert/if/loop/parallel/retry/...)', credits: 0 },
    ],
    ['budget_cents caps spend: a breach stops the run with GUARD_EXCEEDED'],
  );
  assertTotal(
    report,
    estimateWorkflowRunCredits({ taskSteps: args.taskSteps, cuaVersion: args.cuaVersion }),
  );
  return report;
}

export interface MachineArgs {
  osType: OsType;
  hours: number;
  stoppedHours: number;
  snapshots: number;
}

export function machineReport(partial: Partial<MachineArgs> = {}): CostReport {
  const args: MachineArgs = {
    osType: 'linux',
    hours: 1,
    stoppedHours: 0,
    snapshots: 0,
    ...partial,
  };
  const runningRate =
    args.osType === 'windows'
      ? PRICING.machineHourly.windowsRunning
      : PRICING.machineHourly.linuxRunning;
  const items: EstimateLineItem[] = [
    {
      label: `running ${args.osType} ${String(args.hours)}h @ ${String(runningRate)} cr/hr (floored per minute)`,
      credits: estimateMachineRuntimeCredits({ osType: args.osType, minutes: args.hours * 60 }),
    },
  ];
  if (args.stoppedHours > 0) {
    items.push({
      label: `stopped ${String(args.stoppedHours)}h @ ${String(PRICING.machineHourly.stopped)} cr/hr (any OS)`,
      credits: estimateMachineRuntimeCredits({
        osType: args.osType,
        minutes: args.stoppedHours * 60,
        status: 'stopped',
      }),
    });
  }
  if (args.snapshots > 0) {
    items.push({
      label: `snapshots (${String(args.snapshots)} x ${String(snapshotCredits())} cr, refunded on failure)`,
      credits: args.snapshots * snapshotCredits(),
    });
  }
  return finalize(`machine (${args.osType})`, items, [
    `provisioning gate: wallet >= ${String(PRICING.provisioningGateCredits)} cr ($0.20) — a gate, not a fee`,
    'runtime is metered per minute, rounded DOWN; per-call ops (actions/terminal/files/browser/screenshots) are free',
    'a wallet-dry machine is STOPPED (suspended_for_billing), never destroyed',
  ]);
}

function assertTotal(report: CostReport, expected: number): void {
  if (report.totalCredits !== expected) {
    throw new Error(
      `internal pricing mismatch for "${report.title}": itemized ${String(report.totalCredits)} ` +
        `!= canonical ${String(expected)} — pricing table drift, fix cost.ts/ex10 together`,
    );
  }
}

// ---------------------------------------------------------------------------
// Plan mode — total a JSON batch
// ---------------------------------------------------------------------------

export type PlanEntry =
  | ({ kind: 'predict' } & Partial<InferenceArgs>)
  | ({ kind: 'session' } & Partial<SessionArgs>)
  | ({ kind: 'ground' } & Partial<Pick<InferenceArgs, 'width' | 'height'>>)
  | ({ kind: 'run' } & Partial<RunArgs>)
  | ({ kind: 'workflow' } & Partial<{ taskSteps: number; cuaVersion: CuaVersion }>)
  | ({ kind: 'machine' } & Partial<MachineArgs>);

export function reportForEntry(entry: PlanEntry): CostReport {
  switch (entry.kind) {
    case 'predict':
      return predictReport(entry);
    case 'session':
      return sessionReport(entry);
    case 'ground':
      return groundReport(entry);
    case 'run':
      return runReport(entry);
    case 'workflow':
      return workflowReport(entry);
    case 'machine':
      return machineReport(entry);
  }
}

const PLAN_KINDS = ['predict', 'session', 'ground', 'run', 'workflow', 'machine'] as const;

/** Parse + total a JSON batch: `[{"kind":"predict","width":1280,...}, ...]`. */
export function planReportFromJson(jsonText: string): CostReport {
  let parsed: unknown;
  try {
    parsed = JSON.parse(jsonText);
  } catch (cause) {
    throw new UsageError(`plan file is not valid JSON: ${String(cause)}`);
  }
  if (!Array.isArray(parsed)) throw new UsageError('plan must be a JSON ARRAY of entries');
  const items: EstimateLineItem[] = parsed.map((entry: unknown, index) => {
    if (typeof entry !== 'object' || entry === null || Array.isArray(entry)) {
      throw new UsageError(`plan[${String(index)}] must be an object with a "kind"`);
    }
    const kind = (entry as Record<string, unknown>).kind;
    if (typeof kind !== 'string' || !(PLAN_KINDS as readonly string[]).includes(kind)) {
      throw new UsageError(
        `plan[${String(index)}].kind ${JSON.stringify(kind)} must be one of: ${PLAN_KINDS.join(', ')}`,
      );
    }
    const report = reportForEntry(entry as PlanEntry);
    return { label: `[${String(index)}] ${report.title}`, credits: report.totalCredits };
  });
  return finalize(`plan (${String(items.length)} entries)`, items);
}

// ---------------------------------------------------------------------------
// CLI parsing (testable)
// ---------------------------------------------------------------------------

export const USAGE = `usage: npx tsx src/examples/ex10-cost-helper.ts <subcommand> [flags]

subcommands and flags:
  predict   [--width N] [--height N] [--trajectory N] [--cua v1|v3|v4] [--prompt-chars N]
  session   [--steps N] + the predict flags
  ground    [--width N] [--height N]
  run       [--steps N] [--cua v1|v3|v4]
  workflow  [--task-steps N] [--cua v1|v3|v4]
  machine   [--os linux|windows | --windows] [--hours H] [--stopped-hours H] [--snapshots N]
  plan      --file <batch.json>   (JSON array of {kind, ...flags-as-fields})`;

function toNumber(name: string, raw: string, opts: { int?: boolean; min?: number }): number {
  const value = Number(raw);
  if (!Number.isFinite(value)) throw new UsageError(`${name} must be a number (got ${raw})`);
  if (opts.int === true && !Number.isInteger(value)) {
    throw new UsageError(`${name} must be an integer (got ${raw})`);
  }
  if (opts.min !== undefined && value < opts.min) {
    throw new UsageError(`${name} must be >= ${String(opts.min)} (got ${raw})`);
  }
  return value;
}

function toCua(raw: string): CuaVersion {
  if (raw !== 'v1' && raw !== 'v3' && raw !== 'v4') {
    throw new UsageError(`--cua must be v1, v3 or v4 (got ${raw})`);
  }
  return raw;
}

function toCostOptions(args: InferenceArgs): {
  cuaVersion: CuaVersion;
  screenWidth: number;
  screenHeight: number;
  trajectoryScreenshots: number;
  systemPromptChars: number;
} {
  return {
    cuaVersion: args.cuaVersion,
    screenWidth: args.width,
    screenHeight: args.height,
    trajectoryScreenshots: args.trajectory,
    systemPromptChars: args.promptChars,
  };
}

/** Parse argv (subcommand first) into a finished report. Throws UsageError. */
export function buildReportFromArgs(argv: readonly string[]): CostReport {
  const [subcommand, ...rest] = argv;
  if (subcommand === undefined) throw new UsageError('missing subcommand');

  const { values } = parseArgs({
    args: [...rest],
    strict: true,
    allowPositionals: false,
    options: {
      width: { type: 'string' },
      height: { type: 'string' },
      trajectory: { type: 'string' },
      cua: { type: 'string' },
      'prompt-chars': { type: 'string' },
      steps: { type: 'string' },
      'task-steps': { type: 'string' },
      hours: { type: 'string' },
      'stopped-hours': { type: 'string' },
      snapshots: { type: 'string' },
      os: { type: 'string' },
      windows: { type: 'boolean' },
      file: { type: 'string' },
    },
  });

  const inference = (): Partial<InferenceArgs> => ({
    ...(values.width === undefined
      ? {}
      : { width: toNumber('--width', values.width, { int: true, min: 1 }) }),
    ...(values.height === undefined
      ? {}
      : { height: toNumber('--height', values.height, { int: true, min: 1 }) }),
    ...(values.trajectory === undefined
      ? {}
      : { trajectory: toNumber('--trajectory', values.trajectory, { int: true, min: 0 }) }),
    ...(values.cua === undefined ? {} : { cuaVersion: toCua(values.cua) }),
    ...(values['prompt-chars'] === undefined
      ? {}
      : { promptChars: toNumber('--prompt-chars', values['prompt-chars'], { int: true, min: 0 }) }),
  });

  switch (subcommand) {
    case 'predict':
      return predictReport(inference());
    case 'session':
      return sessionReport({
        ...inference(),
        ...(values.steps === undefined
          ? {}
          : { steps: toNumber('--steps', values.steps, { int: true, min: 0 }) }),
      });
    case 'ground':
      return groundReport(inference());
    case 'run':
      return runReport({
        ...(values.steps === undefined
          ? {}
          : { steps: toNumber('--steps', values.steps, { int: true, min: 1 }) }),
        ...(values.cua === undefined ? {} : { cuaVersion: toCua(values.cua) }),
      });
    case 'workflow':
      return workflowReport({
        ...(values['task-steps'] === undefined
          ? {}
          : { taskSteps: toNumber('--task-steps', values['task-steps'], { int: true, min: 0 }) }),
        ...(values.cua === undefined ? {} : { cuaVersion: toCua(values.cua) }),
      });
    case 'machine': {
      const osType: OsType | undefined =
        values.windows === true
          ? 'windows'
          : values.os === undefined
            ? undefined
            : parseOs(values.os);
      return machineReport({
        ...(osType === undefined ? {} : { osType }),
        ...(values.hours === undefined
          ? {}
          : { hours: toNumber('--hours', values.hours, { min: 0 }) }),
        ...(values['stopped-hours'] === undefined
          ? {}
          : { stoppedHours: toNumber('--stopped-hours', values['stopped-hours'], { min: 0 }) }),
        ...(values.snapshots === undefined
          ? {}
          : { snapshots: toNumber('--snapshots', values.snapshots, { int: true, min: 0 }) }),
      });
    }
    case 'plan': {
      if (values.file === undefined) throw new UsageError('plan requires --file <batch.json>');
      return planReportFromJson(readFileSync(values.file, 'utf8'));
    }
    default:
      throw new UsageError(`unknown subcommand "${subcommand}"`);
  }
}

function parseOs(raw: string): OsType {
  if (raw !== 'linux' && raw !== 'windows') {
    throw new UsageError(`--os must be linux or windows (got ${raw})`);
  }
  return raw;
}

export function renderReport(report: CostReport, sandbox: boolean): string {
  const lines = [
    report.title,
    formatEstimate(report.items, { sandbox }),
    ...report.notes.map((note) => `  note: ${note}`),
  ];
  return lines.join('\n');
}

// ---------------------------------------------------------------------------
// Thin CLI
// ---------------------------------------------------------------------------

function isMain(): boolean {
  const entry = process.argv[1];
  return entry !== undefined && path.resolve(entry) === fileURLToPath(import.meta.url);
}

export function main(argv: string[] = process.argv.slice(2)): void {
  const print = (line: string): void => void process.stdout.write(`${line}\n`);
  try {
    const report = buildReportFromArgs(argv);
    print(renderReport(report, isSandboxKey()));
  } catch (error) {
    console.error(error instanceof Error ? error.message : String(error));
    process.exitCode = 1;
  }
}

if (isMain()) main();

File: python/examples/ex10_cost_helper.py
"""Example 10 -- Cost helper CLI: estimate any Coasty operation before you buy.

Purpose
    A purely local calculator over :mod:`coasty.cost` (the full pricing
    table, 1 credit = $0.01). Subcommands estimate single operations;
    ``plan`` reads a small JSON file describing an intended batch and totals
    it, itemized.

Flow
    argparse subcommand -> coasty.cost.estimate_* -> itemized credits + USD.

Subcommands
    predict   --cua-version --width --height --trajectory --system-prompt-chars
    session   --steps + the same per-step surcharge flags (create = 10 cr flat)
    ground    --width --height
    run       --steps --cua-version            (5 cr/step v3+v4, 8 cr v1)
    workflow  --task-steps --cua-version       (control-flow steps are free)
    machine   --os --hours --state --snapshots (hourly, metered per minute)
    plan      PATH.json                        (totals a whole batch)

Endpoints
    None -- entirely offline arithmetic (so no spend gate either).

Estimated cost
    $0.00 to run this example. Remember: sandbox keys (sk-coasty-test-*)
    never bill, so every estimate is $0 in sandbox; and charges are debited
    up front then AUTO-REFUNDED on failure (e.g. PREDICTION_FAILED, a failed
    snapshot), so a failed op costs nothing.

Run
    python examples/ex10_cost_helper.py predict --width 1280 --height 720
    python examples/ex10_cost_helper.py machine --os windows --hours 2
    python examples/ex10_cost_helper.py plan my_batch.json
"""

from __future__ import annotations

import argparse
import json
import sys
from collections.abc import Mapping, Sequence
from pathlib import Path
from typing import Any, cast

from coasty import cost, env
from coasty.cost import CostEstimate, CostItem
from coasty.types import CuaVersion, OsType

SANDBOX_NOTE = "note: sandbox keys (sk-coasty-test-*) never bill -- every estimate is $0.00"
REFUND_NOTE = (
    "note: charges are debited up front and auto-refunded on failure "
    "(failed predictions/groundings/snapshots cost nothing)"
)

PLAN_KINDS = frozenset(
    {
        "predict",
        "session_create",
        "session_predict",
        "ground",
        "parse",
        "run",
        "workflow",
        "machine",
        "snapshot",
    }
)


class PlanError(ValueError):
    """A plan file is malformed (unknown kind, missing field, bad type)."""


def _scaled(estimate: CostEstimate, count: int, unit: str) -> CostEstimate:
    """Multiply an estimate by ``count``, keeping the itemization readable."""
    if count < 0:
        raise ValueError("count must be >= 0")
    if count == 1:
        return estimate
    items = tuple(
        CostItem(f"{item.label} [x{count} {unit}]", item.credits * count) for item in estimate.items
    )
    return CostEstimate(items=items)


def _cua(value: str) -> CuaVersion:
    if value not in ("v1", "v3", "v4"):
        raise argparse.ArgumentTypeError(f"cua_version must be v1/v3/v4, got {value!r}")
    return cast(CuaVersion, value)


def _os(value: str) -> OsType:
    if value not in ("linux", "windows"):
        raise argparse.ArgumentTypeError(f"os must be linux/windows, got {value!r}")
    return cast(OsType, value)


# ── per-subcommand estimators (pure; known-answer tested) ──────────────────


def estimate_predict_cmd(
    *,
    cua_version: CuaVersion = "v3",
    width: int = 1920,
    height: int = 1080,
    trajectory: int = 0,
    system_prompt_chars: int = 0,
    count: int = 1,
) -> CostEstimate:
    """POST /predict: 5 cr + trajectory(+2 ea) + HD(+1 ea) + v1(+3) + prompt(+1)."""
    one = cost.estimate_predict(
        cua_version=cua_version,
        screen_width=width,
        screen_height=height,
        trajectory_sizes=[(width, height)] * trajectory,
        system_prompt="x" * system_prompt_chars if system_prompt_chars > 0 else None,
    )
    return _scaled(one, count, "calls")


def estimate_session_cmd(
    *,
    steps: int = 5,
    cua_version: CuaVersion = "v3",
    width: int = 1920,
    height: int = 1080,
    trajectory_per_step: int = 0,
    system_prompt_chars: int = 0,
) -> CostEstimate:
    """One session: create (10 cr flat, no surcharges) + N predict steps (4 cr +)."""
    if steps < 0:
        raise ValueError("steps must be >= 0")
    step = cost.estimate_session_predict(
        cua_version=cua_version,
        screen_width=width,
        screen_height=height,
        trajectory_sizes=[(width, height)] * trajectory_per_step,
        system_prompt="x" * system_prompt_chars if system_prompt_chars > 0 else None,
    )
    return cost.combine(cost.estimate_session_create(), _scaled(step, steps, "steps"))


def estimate_ground_cmd(*, width: int = 1920, height: int = 1080, count: int = 1) -> CostEstimate:
    """POST /ground: 3 cr, +1 if HD (strictly w>1280 or h>720)."""
    return _scaled(cost.estimate_ground(screen_width=width, screen_height=height), count, "calls")


def estimate_run_cmd(*, steps: int, cua_version: CuaVersion = "v3") -> CostEstimate:
    """A task run: 5 cr per step on v3/v4, 8 cr on v1; no other surcharges."""
    return cost.estimate_run(steps=steps, cua_version=cua_version)


def estimate_workflow_cmd(*, task_steps: int, cua_version: CuaVersion = "v3") -> CostEstimate:
    """A workflow run: only task steps bill; control-flow steps are free."""
    return cost.estimate_workflow_run(task_steps=task_steps, cua_version=cua_version)


def estimate_machine_cmd(
    *,
    os_type: OsType = "linux",
    hours: float = 1.0,
    state: str = "running",
    snapshots: int = 0,
) -> CostEstimate:
    """Machine runtime (hourly rate, metered per minute, rounded down) + snapshots."""
    if hours < 0:
        raise ValueError("hours must be >= 0")
    runtime = cost.estimate_machine_runtime(os_type=os_type, state=state, minutes=hours * 60)
    if snapshots <= 0:
        return runtime
    return cost.combine(runtime, _scaled(cost.estimate_snapshot(), snapshots, "snapshots"))


# ── plan mode: total a whole intended batch from JSON ──────────────────────


def _int_field(item: Mapping[str, Any], key: str, default: int) -> int:
    value = item.get(key, default)
    if isinstance(value, bool) or not isinstance(value, int):
        raise PlanError(f"plan item field {key!r} must be an integer, got {value!r}")
    return value


def _float_field(item: Mapping[str, Any], key: str, default: float) -> float:
    value = item.get(key, default)
    if isinstance(value, bool) or not isinstance(value, int | float):
        raise PlanError(f"plan item field {key!r} must be a number, got {value!r}")
    return float(value)


def _cua_field(item: Mapping[str, Any]) -> CuaVersion:
    value = item.get("cua_version", "v3")
    if value not in ("v1", "v3", "v4"):
        raise PlanError(f"plan item cua_version must be v1/v3/v4, got {value!r}")
    return cast(CuaVersion, value)


def estimate_plan_item(item: Mapping[str, Any]) -> CostEstimate:
    """Estimate one ``{"kind": ..., ...}`` entry of a plan file."""
    kind = item.get("kind")
    if not isinstance(kind, str) or kind not in PLAN_KINDS:
        raise PlanError(f"unknown plan item kind {kind!r}; expected one of {sorted(PLAN_KINDS)}")
    count = _int_field(item, "count", 1)
    width = _int_field(item, "screen_width", 1920)
    height = _int_field(item, "screen_height", 1080)

    if kind == "predict":
        return estimate_predict_cmd(
            cua_version=_cua_field(item),
            width=width,
            height=height,
            trajectory=_int_field(item, "trajectory", 0),
            system_prompt_chars=_int_field(item, "system_prompt_chars", 0),
            count=count,
        )
    if kind == "session_create":
        return _scaled(cost.estimate_session_create(), count, "sessions")
    if kind == "session_predict":
        one = cost.estimate_session_predict(
            cua_version=_cua_field(item),
            screen_width=width,
            screen_height=height,
            trajectory_sizes=[(width, height)] * _int_field(item, "trajectory", 0),
            system_prompt=(
                "x" * _int_field(item, "system_prompt_chars", 0)
                if _int_field(item, "system_prompt_chars", 0) > 0
                else None
            ),
        )
        return _scaled(one, count, "steps")
    if kind == "ground":
        return estimate_ground_cmd(width=width, height=height, count=count)
    if kind == "parse":
        return _scaled(cost.estimate_parse(), count, "calls")
    if kind == "run":
        return estimate_run_cmd(steps=_int_field(item, "steps", 1), cua_version=_cua_field(item))
    if kind == "workflow":
        return estimate_workflow_cmd(
            task_steps=_int_field(item, "task_steps", 1), cua_version=_cua_field(item)
        )
    if kind == "machine":
        os_value = item.get("os", "linux")
        if os_value not in ("linux", "windows"):
            raise PlanError(f"plan item os must be linux/windows, got {os_value!r}")
        state = item.get("state", "running")
        if not isinstance(state, str):
            raise PlanError(f"plan item state must be a string, got {state!r}")
        minutes = _float_field(item, "minutes", _float_field(item, "hours", 1.0) * 60)
        return cost.estimate_machine_runtime(
            os_type=cast(OsType, os_value), state=state, minutes=minutes
        )
    # kind == "snapshot"
    return _scaled(cost.estimate_snapshot(), count, "snapshots")


def estimate_plan(plan: Mapping[str, Any]) -> CostEstimate:
    """Total an entire ``{"items": [...]}`` plan, keeping every line itemized."""
    items = plan.get("items")
    if not isinstance(items, list) or not items:
        raise PlanError('a plan needs a non-empty "items" list')
    estimates = []
    for index, item in enumerate(items):
        if not isinstance(item, Mapping):
            raise PlanError(f"plan items[{index}] must be an object, got {type(item).__name__}")
        estimates.append(estimate_plan_item(item))
    return cost.combine(*estimates)


def load_plan(path: Path) -> dict[str, Any]:
    """Read + parse a plan file (raises PlanError with a clear message)."""
    try:
        raw = path.read_text(encoding="utf-8")
    except OSError as exc:
        raise PlanError(f"cannot read plan file {path}: {exc}") from exc
    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise PlanError(f"plan file {path} is not valid JSON: {exc}") from exc
    if not isinstance(parsed, dict):
        raise PlanError(f"plan file {path} must contain a JSON object")
    return parsed


# ── CLI plumbing ───────────────────────────────────────────────────────────


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__.split("\n", maxsplit=1)[0])
    sub = parser.add_subparsers(dest="command", required=True)

    def add_screen_flags(p: argparse.ArgumentParser) -> None:
        p.add_argument("--width", type=int, default=1920, help="screenshot width (default 1920)")
        p.add_argument("--height", type=int, default=1080, help="screenshot height (default 1080)")

    p_predict = sub.add_parser("predict", help="one or more POST /predict calls")
    p_predict.add_argument("--cua-version", type=_cua, default="v3")
    add_screen_flags(p_predict)
    p_predict.add_argument("--trajectory", type=int, default=0, help="prior screenshots attached")
    p_predict.add_argument("--system-prompt-chars", type=int, default=0)
    p_predict.add_argument("--count", type=int, default=1)

    p_session = sub.add_parser("session", help="a session: create + N predict steps")
    p_session.add_argument("--steps", type=int, default=5)
    p_session.add_argument("--cua-version", type=_cua, default="v3")
    add_screen_flags(p_session)
    p_session.add_argument("--trajectory-per-step", type=int, default=0)
    p_session.add_argument("--system-prompt-chars", type=int, default=0)

    p_ground = sub.add_parser("ground", help="one or more POST /ground calls")
    add_screen_flags(p_ground)
    p_ground.add_argument("--count", type=int, default=1)

    p_run = sub.add_parser("run", help="a task run of N agent steps")
    p_run.add_argument("--steps", type=int, required=True)
    p_run.add_argument("--cua-version", type=_cua, default="v3")

    p_workflow = sub.add_parser("workflow", help="a workflow run of N billed task steps")
    p_workflow.add_argument("--task-steps", type=int, required=True)
    p_workflow.add_argument("--cua-version", type=_cua, default="v3")

    p_machine = sub.add_parser("machine", help="machine runtime + optional snapshots")
    p_machine.add_argument("--os", type=_os, default="linux")
    p_machine.add_argument("--hours", type=float, default=1.0)
    p_machine.add_argument(
        "--state", choices=("running", "stopped", "suspended"), default="running"
    )
    p_machine.add_argument("--snapshots", type=int, default=0)

    p_plan = sub.add_parser("plan", help="total a JSON batch plan")
    p_plan.add_argument("path", type=Path, help='JSON file: {"items": [{"kind": ...}, ...]}')

    return parser


def estimate_from_args(args: argparse.Namespace) -> CostEstimate:
    """Map parsed CLI args onto the right estimator (pure, easily tested)."""
    command = str(args.command)
    if command == "predict":
        return estimate_predict_cmd(
            cua_version=args.cua_version,
            width=args.width,
            height=args.height,
            trajectory=args.trajectory,
            system_prompt_chars=args.system_prompt_chars,
            count=args.count,
        )
    if command == "session":
        return estimate_session_cmd(
            steps=args.steps,
            cua_version=args.cua_version,
            width=args.width,
            height=args.height,
            trajectory_per_step=args.trajectory_per_step,
            system_prompt_chars=args.system_prompt_chars,
        )
    if command == "ground":
        return estimate_ground_cmd(width=args.width, height=args.height, count=args.count)
    if command == "run":
        return estimate_run_cmd(steps=args.steps, cua_version=args.cua_version)
    if command == "workflow":
        return estimate_workflow_cmd(task_steps=args.task_steps, cua_version=args.cua_version)
    if command == "machine":
        return estimate_machine_cmd(
            os_type=args.os, hours=args.hours, state=args.state, snapshots=args.snapshots
        )
    if command == "plan":
        return estimate_plan(load_plan(args.path))
    raise ValueError(f"unknown command {command!r}")  # unreachable via argparse


def main(argv: Sequence[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    try:
        estimate = estimate_from_args(args)
    except (PlanError, ValueError) as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 2

    api_key = env.get_api_key()
    sandbox = api_key is not None and env.is_sandbox_key(api_key)
    title = f"Estimated cost ({args.command})"
    print(cost.format_estimate(estimate, title=title, sandbox=sandbox))
    if not sandbox:
        print(SANDBOX_NOTE)
    print(REFUND_NOTE)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
